csi_camera.cpp

```diff
@@ -0,0 +1,451 @@
+            }
+
+            return image;
+        }
+
+        API dynamic_api() const override {
+            return Camera::static_api();
+        }
+
+        AttributeMap do_command(AttributeMap command) override {
+            std::cerr << "do_command not implemented" << std::endl;
+            return 0;
+        }
+
+        point_cloud get_point_cloud(std::string mime_type) override {
+            std::cerr << "get_point_cloud not implemented" << std::endl;
+            return point_cloud{};
+        }
+        std::vector<GeometryConfig> get_geometries() override {
+            std::cerr << "get_geometries not implemented" << std::endl;
+            return std::vector<GeometryConfig>{};
+        }
+
+        properties get_properties() override {
+            std::cerr << "get_properties not implemented" << std::endl;
+            return properties{};
+        };
+
+        // GST
+
+        void init_csi(std::string pipeline_args) {
+            // Build gst pipeline
+            pipeline = gst_parse_launch(
+                pipeline_args.c_str(),
+                nullptr
+            );
+            if (!pipeline) {
+                std::cerr << "Failed to create the pipeline" << std::endl;
+                std::exit(EXIT_FAILURE);
+            }
+
+            // Print pipeline structure
+            if (debug) {
+                GST_DEBUG_BIN_TO_DOT_FILE(GST_BIN(pipeline), GST_DEBUG_GRAPH_SHOW_ALL, "pipeline_structure");
+            }
+            
+            // Fetch the appsink element
+            appsink = gst_bin_get_by_name(GST_BIN(pipeline), "appsink0");
+            if (!appsink) {
+                std::cerr << "Failed to get the appsink element" << std::endl;
+                gst_object_unref(pipeline);
+                std::exit(EXIT_FAILURE);
+            }
+
+            // Start the pipeline
+            if (gst_element_set_state(pipeline, GST_STATE_PLAYING) == GST_STATE_CHANGE_FAILURE) {
+                std::cerr << "Failed to start the pipeline" << std::endl;
+                gst_object_unref(appsink);
+                gst_object_unref(pipeline);
+                std::exit(EXIT_FAILURE);
+            }
+
+            // Handle async pipeline creation
+            wait_pipeline();
+
+            // Run the main loop
+            bus = gst_element_get_bus(pipeline);
+            if (!bus) {
+                std::cerr << "Failed to get the bus for the pipeline" << std::endl;
+                gst_object_unref(appsink);
+                gst_object_unref(pipeline);
+                std::exit(EXIT_FAILURE);
+            }
+
+            return;
+        }
+
+        // Handles async GST state change
+        void wait_pipeline() {
+            GstState state, pending;
+            GstStateChangeReturn ret;
+            while ((ret = gst_element_get_state(pipeline, &state, &pending, GST_CLOCK_TIME_NONE)) == GST_STATE_CHANGE_ASYNC) {
+                // wait for state change to complete
```

Comment:
Sleep so we don't busy loop?

Reply:
Yes this loop simply hangs execution until state change occurs in the GST pipeline. 

I did not want to do this in the background as it is much cleaner to handle errors/exits synchronously.

Reply:
We don't need to do this in the background, but we should still sleep in the body of this while loop. We should also add a timeout, so we don't wait indefinitely.

Reply:
Great point! Added 1 second timeout.

Reply:
ok I see what you saying.

Added 5 second timeout to busy loop. Also waiting 10 ms between state queries to avoid using up too much cpu.

Reply:
Thank you!